[SVLS-9486] Add AWS Lambda Durable Functions documentation (#38166)
DataDog/documentation: content/en/serverless/aws_lambda/durable_functions.md

---

title: Monitoring AWS Lambda Durable Functions
description: Set up Datadog Serverless Monitoring for AWS Lambda Durable Functions.

---

Datadog provides full visibility into the metrics, logs, and traces for AWS Lambda Durable Function executions. In a single view, you can monitor your AWS Lambda Durable Functions alongside your other serverless compute services to spot bottlenecks and fix errors.

**Supported runtimes:** Python, Node.js

## Setup

### Instrument your Lambda function

1. Follow [the steps for instrumenting a Lambda function][1]. Make sure the Datadog Lambda Library is installed and tracing is enabled. Use at least the following versions:

    - Datadog Lambda Extension: v98+
    - Datadog Python Lambda layer: v126+
    - Datadog Node.js Lambda layer: v141+

2. Set the following environment variable on your Lambda function:

    ```text

>     DD_LAMBDA_DURABLE_FUNCTION_LOG_BUFFER_SIZE=5

Reviewer: Can we add a brief clause explaining what this value controls (e.g., number of buffered log lines) and its default/valid range?

Author: Added.
By the way, I plan to change the default from 0 to 5 later to simplify onboarding, but I guess I shouldn't mention this plan yet as it's just a plan.

   [![Launch Stack](https://s3.amazonaws.com/cloudformation-examples/cloudformation-launch-stack.png)][2]

2. Provide your Datadog API key using exactly one of the following parameters:

    - `DdApiKey`: your [Datadog API key][3] in plaintext
    - `DdApiKeySecretArn`: the ARN of an AWS Secrets Manager secret whose value is your Datadog API key
    - `DdApiKeySsmParameterName`: the name of an AWS Systems Manager Parameter Store `SecureString` parameter that holds your Datadog API key

3. Set `DdSite` to your [Datadog site][4]. The default is `datadoghq.com`.
4. Optionally, restrict which events are forwarded:

    - `Statuses`: a comma-separated list of execution statuses to forward. Valid values are `RUNNING`, `SUCCEEDED`, `FAILED`, `TIMED_OUT`, and `STOPPED`. Leave this empty to forward all statuses.
    - `FunctionArnFilter1` through `FunctionArnFilter5`: up to five unqualified Lambda function ARNs or EventBridge wildcard patterns, such as `arn:aws:lambda:us-east-2:123456789012:function:my-durable-*`. Do not include a version or alias suffix. Leave all five empty to forward events from every function in the region.

>     - `BufferIntervalSeconds`: the Amazon Data Firehose buffer interval, in seconds (`60`-`900`, default `60`). Higher values reduce the number of outbound requests at the cost of freshness.

Reviewer: ```suggestion
    - `BufferIntervalSeconds`: the Amazon Data Firehose buffer interval, in seconds (`60`–`900`, default `60`). Higher values reduce the number of outbound requests at the cost of freshness.
```

Author: Fixed

5. Click **Create stack** and wait for the stack to finish creating.
6. If your durable functions run in multiple AWS regions, repeat these steps in each region.

> Install the [AWS Lambda integration][5] in Datadog. Its out-of-the-box logs pipeline transforms the forwarded events.

Reviewer: Seems odd to have an instruction line separate from the numbered steps.  Either fold this in as step 7, or give it the same subsection treatment as the "Create a trace retention filter" step, for consistency.

Author: This is part of "Forward durable execution events", so I made both this line and "Deploy the CloudFormation stack" bullet points. Does it look good to you?

---

title: Monitoring AWS Lambda Durable Functions
description: Install and configure the Datadog Agent for AWS Lambda Durable Functions.

---

> Datadog provides full visibility into the metrics, logs, and traces for AWS Lambda Durable Function executions. You can monitor your AWS Lambda Durable Functions alongside your other serverless compute services in a single, unified view. This enables you to spot bottlenecks and fix errors effectively.

Reviewer: ```suggestion
Datadog provides full visibility into the metrics, logs, and traces for AWS Lambda Durable Function executions. In a single view, you can monitor your AWS Lambda Durable Functions alongside your other serverless compute services to spot bottlenecks and fix errors.
```

<div class="alert alert-info">
<strong>Supported runtimes</strong>: Python, Node.js

> </div>

Reviewer: I think formatting it in this way makes it seem like less of an aside... Perhaps just personal preference though. There's no real precedent from what I can tell looking through the other pages in this section.

```suggestion
**Supported runtimes:** Python, Node.js
```

## Setup

### Instrument your Lambda function

> 1. Complete [the common steps for instrumenting a Lambda function][1]. Make sure the Datadog Lambda Library is installed and tracing is enabled, using at least the following versions:

Reviewer: ```suggestion
1. Follow [the steps for instrumenting a Lambda function][1]. Make sure the Datadog Lambda Library is installed and tracing is enabled. Use at least the following versions:
```

    - Datadog Lambda Extension: v98+
    - Datadog Python Lambda layer: v126+
    - Datadog Node.js Lambda layer: v141+

2. Set the following environment variable on your Lambda function:

    ```text
    DD_LAMBDA_DURABLE_FUNCTION_LOG_BUFFER_SIZE=5
    ```

### Forward durable execution events

Deploy the CloudFormation stack to your AWS account to forward durable execution status change events to Datadog:

[![Launch Stack](https://s3.amazonaws.com/cloudformation-examples/cloudformation-launch-stack.png)][2]

> 1. Log in to the AWS account and region where your durable functions run, then open the stack created by the button above.

Reviewer: Folding the button directly into the step that reference it.

```suggestion
1. Log in to the AWS account and region where your durable functions run, then click **Launch Stack** to open the stack:

   [![Launch Stack](https://s3.amazonaws.com/cloudformation-examples/cloudformation-launch-stack.png)][2]
```

2. Provide your Datadog API key using exactly one of the following parameters:

    - `DdApiKey`: your [Datadog API key][3] in plaintext.
    - `DdApiKeySecretArn`: the ARN of an AWS Secrets Manager secret whose value is your Datadog API key.

>     - `DdApiKeySsmParameterName`: the name of an AWS Systems Manager Parameter Store `SecureString` parameter that holds your Datadog API key.

Reviewer: Fragments don't need to end in periods.

```suggestion
    - `DdApiKey`: your [Datadog API key][3] in plaintext
    - `DdApiKeySecretArn`: the ARN of an AWS Secrets Manager secret whose value is your Datadog API key
    - `DdApiKeySsmParameterName`: the name of an AWS Systems Manager Parameter Store `SecureString` parameter that holds your Datadog API key
```

        [![Launch Stack](https://s3.amazonaws.com/cloudformation-examples/cloudformation-launch-stack.png)][2]

    2. Provide your Datadog API key using exactly one of the following parameters:

        - `DdApiKey`: your [Datadog API key][3] in plaintext
        - `DdApiKeySecretArn`: the ARN of an AWS Secrets Manager secret whose value is your Datadog API key
        - `DdApiKeySsmParameterName`: the name of an AWS Systems Manager Parameter Store `SecureString` parameter that holds your Datadog API key

    3. Set `DdSite` to your [Datadog site][4]. The default is `datadoghq.com`.
    4. Optionally, restrict which events are forwarded:

        - `Statuses`: a comma-separated list of execution statuses to forward. Valid values are `RUNNING`, `SUCCEEDED`, `FAILED`, `TIMED_OUT`, and `STOPPED`. Leave this empty to forward all statuses.
        - `FunctionArnFilter1` through `FunctionArnFilter5`: up to five unqualified Lambda function ARNs or EventBridge wildcard patterns, such as `arn:aws:lambda:us-east-2:123456789012:function:my-durable-*`. Do not include a version or alias suffix. Leave all five empty to forward events from every function in the region.

>         - `BufferIntervalSeconds`: the Amazon Data Firehose buffer interval, in seconds (`60`–`900`, default `60`). Higher values reduce the number of outbound requests at the cost of freshness.

Reviewer: ```suggestion
        - `BufferIntervalSeconds`: the Amazon Data Firehose buffer interval, in seconds (`60`-`900`, default `60`). Higher values reduce the number of outbound requests at the cost of freshness.
```
